R/makeNhoods.R

```diff
@@ -48,46 +49,57 @@
         }
-        graph <- graph(x)
-        X_reduced_dims  <- reducedDim(x, reduced_dims)
+        X_graph <- miloR::graph(x)
+        X_reduced_dims  <- SingleCellExperiment::reducedDim(x, reduced_dims)
         if (d > ncol(X_reduced_dims)) {
             warning("Specified d is higher than the total number of dimensions in reducedDim(x, reduced_dims). Falling back to using",
                     ncol(X_reduced_dims),"dimensions\n")
             d <- ncol(X_reduced_dims)
         }
         X_reduced_dims  <- X_reduced_dims[,seq_len(d)]
     } else if(is(x, "igraph")){
-        if(!is.matrix(reduced_dims) & isTRUE(refined)){
+        if(!is.matrix(reduced_dims) & (refined != "none")){
             stop("No reduced dimensions matrix provided - required for refined sampling")
         }
-        graph <- x
+        X_graph <- x
         X_reduced_dims <- reduced_dims
     } else{
         stop("Data format: ", class(x), " not recognised. Should be Milo or igraph")
     }
-    random_vertices <- .sample_vertices(graph, prop, return.vertices = TRUE)
+    random_vertices <- .sample_vertices(X_graph, prop, return.vertices = TRUE, seed = seed)
 
-    if (isFALSE(refined)) {
-        sampled_vertices <- random_vertices
-    } else if (isTRUE(refined)) {
+    if(refined == "refined"){
         sampled_vertices <- .refined_sampling(random_vertices, X_reduced_dims, k)
+    } else if (refined == "graph-refined-a") {
+        if(!is.directed(X_graph)){
+            warning("When using graph-refined-a mode, kNN graph must be directed.")
         }
-
-    sampled_vertices <- unique(sampled_vertices)
-
-    nh_mat <- Matrix(data = 0, nrow=ncol(x), ncol=length(sampled_vertices), sparse = TRUE)
-    # Is there an alternative to using a for loop to populate the sparseMatrix here?
-    for (X in seq_len(length(sampled_vertices))){
-        nh_mat[as_ids(neighbors(graph, v = sampled_vertices[X])), X] <- 1
+        sampled_vertices <- .graph_independent_sampling(random_vertices, X_graph)
+    } else if (refined == "graph-refined-b") {
+        X_graph <- igraph::as.undirected(X_graph, mode = "collapse")
+        sampled_vertices <- .graph_independent_sampling(random_vertices, X_graph)
+    } else if (refined == "random") {
+        sampled_vertices <- random_vertices
+    } else {
+        stop("Refined must be one of refined, graph-refined, or random")
     }
-
+    
+    sampled_vertices <- unique(sampled_vertices)
+    
+    #Q: Is there an alternative to using a for loop to populate the sparseMatrix here?
+    #A: https://stackoverflow.com/questions/4942361/how-to-turn-a-list-of-lists-to-a-sparse-matrix-in-r-without-using-lapply 
+    neighbor_list <- lapply(seq_along(sampled_vertices), function(i) as_ids(neighbors(X_graph, v = sampled_vertices[i], mode = "all")))
+    n.ids <- sapply(neighbor_list,length)
+    vals <- unlist(neighbor_list)
+    nh_mat <- Matrix::sparseMatrix(vals, rep(seq_along(n.ids), n.ids), x = 1)
```

Comment:
Is this faster and more memory efficient than the for loop?

Reply:
addressed in vignette below